docs: basic overview of git workflow (#31)
ThinkFiveable/open: git-tutorials/git.md

- [Getting Started with Git/GitHub](#getting-started-with-git-github)
    + [Clone](#clone)
    + [Pull](#pull)
    + [Commit](#commit)
    + [Push](#push)
    + [Pull Request](#pull-request)

# Getting Started with Git/GitHub

> To use Git, you'll need to download it [here](https://git-scm.com/downloads).

Reviewer: ```suggestion
To use Git, you'll need to download it [here](https://git-scm.com/downloads).

Be sure to set your username and email before making contributions.
* `git config --global user.name "Your Name"`
* `git config --global user.email "youremail@yourdomain.com"`

### Clone
First things first, we'll want to get a copy of our repository on our computer locally by cloning our repository. We will need the HTTPS address on the repository page which you can copy/paste from the browser. We will now use our terminal's shell commands to move to the place where we want to put our directory copy.

For example, if we wanted to change to our Documents directory, we would do the following:
`cd Documents`

To clone our repository, we use the following command in our terminal:
`git clone <https address here>`

For example:

> `git clone https://github.com/ThinkFiveable/open fiveable-open`

Reviewer: ```suggestion
`git clone https://github.com/YOURUSERNAMEHERE/open.git fiveable-open`
```

Edit: I originally had this edit change it to: `git clone https://github.com/ThinkFiveable/open.git fiveable-open`, but then I wrote https://github.com/ThinkFiveable/open/pull/31#issuecomment-808788686

Again, this will create a local copy of the repository at the directory specified earlier. You will then need to open these files in your preferred text editor to make your changes.
 
### Pull
You will need to use `git pull` to fetch the latest changes from another repository or local branch. When you run `git pull` in your terminal, remote content will be downloaded and immediately change the local state of your repository to match the new content. It is strongly recommended to not `git pull` when you have uncommitted changes. 

In the event that you do have uncommitted changes and want to `git pull`, you will need to do the following in your command line.

> First, `git stash` which "stashes" your changes away temporarily. Then, `git pull` to fetch the latest changes. Next, `git stash pop` to apply the changes back onto your local copy.

Reviewer: ```suggestion
First, `git stash` which "stashes" your changes away temporarily. Then, `git pull` to fetch the latest changes. Next, `git stash apply` to apply the changes back onto your local copy, and then `git stash drop` to remove those changes from the stash stack.
```
https://codingkilledthecat.wordpress.com/2012/04/27/git-stash-pop-considered-harmful/.

In short, unexpected behavior can occur in the event of pull/merge conflicts, which means git stash pop won't actually remove from the stack like expected. IMO it'd be better to be consistent to begin with than to have to teach edge cases like that.

### Commit
After making your changes, you will use terminal commands to commit them. Our commit message needs to explain what we've done. This message may be useful if we want to check our change history in the future, and it can also be useful to other contributors.

> The syntax for a commit message is as follows `git commit -m "your commit message here"`.

Reviewer: ```suggestion
The syntax for a commit message is as follows `git commit -m "your commit message here"`.
For our repository, please ensure your commit messages follow [this standard](https://github.com/ThinkFiveable/open/blob/main/CONTRIBUTING.md#commiting)
```

    + [Fork](#fork)
    + [Clone](#clone)
    + [Checkout](#checkout)
    + [Pull](#pull)
    + [Commit](#commit)
    + [Push](#push)
    + [Pull Request](#pull-request)
    + [Merge Conflicts](#merge-conflicts)

# Getting Started with Git/GitHub
To use Git, you'll need to download it [here](https://git-scm.com/downloads).

Be sure to set your username and email before making contributions.
In your terminal, do the following:
* `git config --global user.name "Your Name"`
* `git config --global user.email "youremail@yourdomain.com"`

### Fork
First, we will need our own copy of the repository to make changes on. You will need to create a fork by clicking on the `Fork` button in the top right.

### Clone
Next, we'll want to get a copy of our repository on our computer locally by cloning our repository. We will need the HTTPS address on the repository page which you can copy/paste from the browser. We will now use our terminal's shell commands to move to the place where we want to put our directory copy.

For example, if we wanted to change to our Documents directory, we would do the following:
`cd Documents`

To clone our repository, we use the following command in our terminal:

> `git clone <https address here>`

Reviewer: ```suggestion
`git clone <https address here> [cloned-directory-name]`
```

Sounds like a lot right? It's better to prevent all of this from happening in the beginning by just making changes to a new branch every time. So say I want to add a bunch of badges to the readme, I'd make a branch `chore/readme-branch` or `chore/patch-1.` 

To create a new branch from your terminal, do the following:
* `git branch <branchname>` to create a new branch
* `git checkout <branchname>` to swtich to your branch

If you make a bunch of commits and forget to switch branches, you can use `git switch <branchname>` or `git switch -c <branchname>` which will switch to another branch and move your already-done commits to this new branch. Appending `-c` will create a new branch while `git switch <branchname>` switches to an existing branch.
 
### Pull
You will need to use `git pull` to fetch the latest changes from another repository or local branch. When you run `git pull` in your terminal, remote content will be downloaded and immediately change the local state of your repository to match the new content. It is strongly recommended to not `git pull` when you have uncommitted changes. 

> In the event that you do have uncommitted changes and want to `git pull`, you will need to do the following in your command line. First, `git stash` which "stashes" your changes away temporarily. Then, `git pull` to fetch the latest changes. Next, `git stash apply` to apply the changes back onto your local copy, and then `git stash drop` to remove those changes from the stash stack. This method is not reccommened as unexpected behavior may occur.

Reviewer: ```suggestion
In the event that you do have uncommitted changes and want to `git pull`, you will need to do the following in your command line. First, `git stash` which "stashes" your changes away temporarily. Then, `git pull` to fetch the latest changes. Next, `git stash apply` to apply the changes back onto your local copy, and then `git stash drop` to remove those changes from the stash stack. 
```
💀 I think there was a mixup, `git stash pop` is what is not recommended and what'll cause unexpected behavior. `git stash apply` & `git stash drop` are the alternative.

Sounds like a lot right? It's better to prevent all of this from happening in the beginning by just making changes to a new branch every time. So say I want to add a bunch of badges to the readme, I'd make a branch `chore/readme-branch` or `chore/patch-1.` 

To create a new branch from your terminal, do the following:
* `git branch <branchname>` to create a new branch
* `git checkout <branchname>` to swtich to your branch

If you make a bunch of commits and forget to switch branches, you can use `git switch <branchname>` or `git switch -c <branchname>` which will switch to another branch and move your already-done commits to this new branch. Appending `-c` will create a new branch while `git switch <branchname>` switches to an existing branch.
 
### Pull
You will need to use `git pull` to fetch the latest changes from another repository or local branch. When you run `git pull` in your terminal, remote content will be downloaded and immediately change the local state of your repository to match the new content. It is strongly recommended to not `git pull` when you have uncommitted changes. 

> In the event that you do have uncommitted changes and want to `git pull`, you will need to do the following in your command line. First, `git stash` which "stashes" your changes away temporarily. Then, `git pull` to fetch the latest changes. Next, `git stash apply` to apply the changes back onto your local copy, and then `git stash drop` to remove those changes from the stash stack. 

Reviewer 2: why don't you just do git stash pop??

Reviewer: https://github.com/ThinkFiveable/open/pull/31#discussion_r602780056

<img align="center" src="https://cloud-g86tob3jb-hack-club-bot.vercel.app/1screen_shot_2021-03-26_at_9.26.14_pm.png" alt="compare and pull request" width=500/>

As your PR waits to be merged, other peoples' PRs might be merged before yours, meaning you have to "update your branch" with the master branch. You can do it with either this button 

<img align="center" src="https://cdn.discordapp.com/attachments/768856355255484469/825476408767545392/chrome_2e0DN4u2iQ.png" alt="Update branch button"/>

or (the more preferred method as the above button won't always be avaliable), you can do these steps:
- Add our repository as a remote URL (`git remote add upstream https://github.com/ThinkFiveable/open.git`)
- Make sure you are on the branch you want to merge master into (`git checkout PR_BRANCH_NAME`)
- Make sure you have the latest upstream changes (`git fetch upstream`)

> - Merge the upstream master branch into your PR branch (`git merge upstream/master`). At this stage you might run into what are called **MERGE CONFLICTS. You must resolve these before pushing to your fork.** See: [Merge Conflicts](#merge-conflicts)

Author: Should it be `git merge upstream/main`?

Reviewer: Yeah totally forgot we use main
